### PR TITLE
[caffe2] Remove IValue include from operator.h

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #67096
* __->__ #67094

ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

### CODE REVIEW

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

File: caffe2/core/IValueInterface.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  size_t size() const {
+    return size_;
+  }
+  bool empty() const {
+    return size() == 0;
+  }
+
+  const c10::IValue& at(size_t idx) const;
+  std::vector<caffe2::Tensor> toTensorVector(size_t idx) const;
```
Comment (dagitses) [inline caffe2/core/IValueInterface.h:33]: please document the preconditions of these functions, e.g.
`// Expects(isTensorList(idx))` for `toTensorVector`, presumably

**Review by dagitses — APPROVED** [review]

File: caffe2/core/IValueInterface.cc [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+int IValueInterface::compute_input_size() const {
+  if (empty()) {
+    return 0;
+  }
+  if (values_[0].isTensorList()) {
+    // if the first input is a tensor list, we get input tensors by indexing
+    // into that list. currently, this means that only tensors from that list
+    // are accessible as inputs. any hypothetical input tensors that come after
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply caffe2/core/IValueInterface.cc:62]: The logic here comes straight from the existing translation code,  so I presume it's needed. Caffe2 also isn't being actively developed so I wouldn't feel comfortable changing any core functionality.

File: caffe2/core/IValueInterface.cc [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+caffe2::Tensor contiguous(caffe2::Tensor t) {
+  return caffe2::Tensor(at::Tensor(std::move(t)).contiguous());
+}
+
+
+IValueInterface::IValueInterface(c10::ArrayRef<c10::IValue> values):
+  size_(values.size()) {
+  auto values_tmp = std::make_unique<c10::IValue[]>(size_);
+  std::copy_n(values.data(), size_, values_tmp.get());
+  values_ = values_tmp.release();
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply caffe2/core/IValueInterface.cc:21]: I gave it a try and you can't even call `size()` on a vector of incomplete type. I just don't think `std::vector` was intended to support incomplete types.

File: caffe2/core/IValueInterface.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  // These template functions are not defined in the header,
+  // but are explicitly instantiated in IValueInterface.cc
+  template <typename T>
+  typename c10::detail::ivalue_to_const_ref_overload_return<T>::type to(size_t idx) const;
+
+  template <typename T>
+  std::vector<T> toVec(size_t idx) const;
+
+private:
+  c10::IValue *values_;
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply caffe2/core/IValueInterface.h:47]: You mean why is it an array and not just a single `IValue`?  It means we only need one memory allocation, and allows things like `compute_input_size` to happen in one function without repeatedly passing in and out of the interface.

File: caffe2/core/IValueInterface.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (dagitses) [reply caffe2/core/IValueInterface.cc:21]: OK how about defining `values_` as `std::unique_ptr<c10::IValue[]>`?

File: caffe2/core/IValueInterface.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (dagitses) [reply caffe2/core/IValueInterface.cc:62]: sounds good to me.

File: caffe2/core/IValueInterface.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply caffe2/core/IValueInterface.cc:21]: I get the following error:
```
/usr/include/c++/9/bits/unique_ptr.h: In instantiation of ‘typename std::enable_if<std::is_convertible<_Up (*)[], _Tp (*)[]>::value>::type std::default_delete<_Tp []>::operator()(_Up*) const [with _Up = c10::IValue; _Tp = c10::IValue; typename std::enable_if<std::is_convertible<_Up (*)[], _Tp (*)[]>::value>::type = void]’:
/usr/include/c++/9/bits/unique_ptr.h:559:17:   required from ‘std::unique_ptr<_Tp [], _Dp>::~unique_ptr() [with _Tp = c10::IValue; _Dp = std::default_delete<c10::IValue []>]’
/home/peter/git/pytorch/caffe2/core/IValueInterface.h:24:47:   required from here
/usr/include/c++/9/bits/unique_ptr.h:113:16: error: invalid application of ‘sizeof’ to incomplete type ‘c10::IValue’
  113 |  static_assert(sizeof(_Tp)>0,
```

**Comment by peterbell10** [comment]
@dagitses PTAL

File: caffe2/core/IValueInterface.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+namespace c10 {
+struct IValue;
+}
+
+namespace caffe2 {
+namespace detail {
+
+TORCH_API caffe2::Tensor contiguous(caffe2::Tensor);
+
+class TORCH_API IValueInterface {
```
[the remark this answers is not in the document]
Comment (dagitses) [reply caffe2/core/IValueInterface.h:22]: reminder about this comment.

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

File: caffe2/core/IValueInterface.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply caffe2/core/IValueInterface.h:22]: https://github.com/pytorch/pytorch/blob/e701cd6ebc11aa9171d32bd1626e938b165f60c9/caffe2/core/IValueInterface.h#L16-L20

[the remark this answers is not in the document]
Comment (dagitses) [reply caffe2/core/IValueInterface.h:22]: thanks. sorry was looking at wrong commit.

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

File: caffe2/core/IValueInterface.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+/* This class represents input arguments to a caffe2 operator when
+ * called from c10. It's purpose is to act as a compilation barrier
+ * between caffe2 and ATen, so that caffe2 operators don't depend on
+ * ATen operators implicitly through c10::IValue.
+ */
+
+class TORCH_API IValueInterface {
+public:
+  IValueInterface(): values_(nullptr), size_(0) {}
+  IValueInterface(c10::ArrayRef<c10::IValue> values);
```
Comment (dagitses) [inline caffe2/core/IValueInterface.h:25]: are you able to mark this explicit?

Comment (peterbell10) [reply caffe2/core/IValueInterface.h]: done

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

**Pull request closed without merging**